contracts/src/StakeTable.sol

```diff
@@ -52,493 +102,356 @@ contract StakeTable is AbstractStakeTable, Ownable, InitializedAt {
-    /// TODO: add epoch logic so that we can ensure the node has first requested to exit and waiting
-    /// for the exit escrow period to be over
-    function withdrawFunds() external virtual override returns (uint256) {
-        Node memory node = nodes[msg.sender];
-
-        if (node.account == address(0)) {
-            revert NodeNotRegistered();
-        }
+        emit Withdrawal(msg.sender, amount);
+    }
 
-        if (node.account != msg.sender) {
-            revert Unauthenticated();
+    /// @notice Withdraw previously delegated funds after a validator has exited
+    /// @param validator The validator to withdraw from
+    function claimValidatorExit(address validator) external virtual {
+        uint256 unlocksAt = validatorExits[msg.sender];
+        if (unlocksAt == 0) {
+            revert ValidatorNotExited();
         }
 
-        uint256 balance = node.balance;
-        if (balance == 0) {
-            // then there's nothing to withdraw but revert so that they're aware that the withdrawal
-            // failed
-            revert InsufficientStakeBalance(0);
+        if (block.timestamp < unlocksAt) {
+            revert PrematureWithdrawal();
         }
 
-        // Verify that the exit escrow period is over.
-        if (currentEpoch() < node.exitEpoch + exitEscrowPeriod(node)) {
-            revert PrematureWithdrawal();
+        uint256 amount = delegations[validator][msg.sender];
+        if (amount == 0) {
+            revert NothingToWithdraw();
         }
-        totalStake -= balance;
 
-        // Delete the node from the stake table.
-        delete nodes[msg.sender];
+        // Mark funds as spent
+        delegations[validator][msg.sender] = 0;
 
-        // Transfer the balance to the node's account.
-        SafeTransferLib.safeTransfer(ERC20(tokenAddress), node.account, balance);
+        SafeTransferLib.safeTransfer(token, msg.sender, amount);
 
-        return balance;
+        emit Withdrawal(msg.sender, amount);
     }
 
     /// @notice Update the consensus keys for a validator
     /// @dev This function is used to update the consensus keys for a validator
-    /// @dev This function can only be called by the validator itself when it's not in the exit
-    /// queue
+    /// @dev This function can only be called by the validator itself when it hasn't exited
+    ///      TODO: MA: is this a good idea? Why should key rotation be blocked for an exiting
+    ///      validator?
     /// @dev The validator will need to give up either its old BLS key and/or old Schnorr key
     /// @dev The validator will need to provide a BLS signature to prove that the account owns the
     /// new BLS key
     /// @param newBlsVK The new BLS verification key
     /// @param newSchnorrVK The new Schnorr verification key
     /// @param newBlsSig The BLS signature that the account owns the new BLS key
+    ///
+    /// TODO: MA: I think this function should be reworked. Is it fine to always force updating both
```

Comment:
i believe the key generation tool generates both a schnorr and blskey at the same time - so you might as well update both? but i could see reasons for not doing so
cc @rob-maron